Make range checks in semConv

GULPF opened the pull request:
Another compiler adventure

Previously the type conversion range checks where only added by `transf.nim`, and evaluated statically when possible. This method missed the cases where the type conversion had already been done during semchecking due to constant folding. To solve this I've changed `checkConvertible` to also make sure that the value is within the range of the destination type.  

I'm not entirely happy with this solution. I feel it should be possible to let `changeType` do the range checking, but I'm not sure how `semConv` would need to change to allow for that.

This new semcheck found a bug in `posix` that I've corrected. It also found a bug in the [opengl package](https://github.com/nim-lang/opengl) which causes a test failure, but I don't know how to fix that one. Relevant line:

https://github.com/nim-lang/opengl/blob/4106d6efc6ad32d848fe75d89f629c8a97012940/src/opengl/private/constants.nim#L4692

`0xFFFFFFFFFFFFFFFF` obviously doesn't fit into `GLenum` (an alias for unit32), so I'm not sure what to do.

Fixes #7102
Fixes #7580
Fixes #6679

GULPF commented:
AppVeyor failure indicates a bug in `winlean.nim`: 
```nim
const
 IOC_OUT* = 0x40000000
 IOC_IN*  = 0x80000000
 IOC_WS2* = 0x08000000
 IOC_INOUT* = IOC_IN or IOC_OUT

template WSAIORW*(x,y): untyped = (IOC_INOUT or x or y)

const
  SIO_GET_EXTENSION_FUNCTION_POINTER* = WSAIORW(IOC_WS2,6).DWORD
```
I think `IOC_IN` is supposed to be a `int32` literal, but the compiler decides that it's `int64`, which results in a different value.

GULPF commented:
Previous comment is incorrect. It appears that [DWORD is supposed to be uin32](https://msdn.microsoft.com/en-us/library/windows/desktop/aa383751(v=vs.85).aspx), not int32 like `winlean` defines. `0x80000000` fits inside uint32. Several other `winlean` types are also incorrect.

Araq commented:
This is pretty important, please fix the CI builds and continue to work on this.

GULPF commented:
OK, I've cleaned up the PR a bit, fixed the winlean constants and submitted https://github.com/nim-lang/opengl/pull/51 which fixes the other test failure.

I haven't changed `DWORD` to `uint32` because that's a big breaking change which is unrelated to this PR, and I don't know if it's correct anyway.

cooldome commented:
Supertb, absolute must have.

Araq, looks like DWORD needs to be changed to uint32 what do you think?

GULPF commented:
After getting more familiar with semfold.nim, I've tried another approach for this that I think is much better. I haven't really tested it yet and it's not finished, but it looks good.

GULPF commented:
Turns out `semfold` isn't applied in a lot of cases (e.g `let i = Natural(-1)` is folded by the `transf` pass, not by `semfold`), so the original idea is more robust after all.

GULPF commented:
The last failing test is because the tests doesn't use the devel version of https://github.com/nim-lang/opengl/. The opengl package should probably get a version bump.

GULPF commented on compiler/semexprs.nim:
```diff
@@ -115,15 +117,23 @@ proc checkConvertible(c: PContext, castDest, src: PType): TConvStatus =
+        result = convNotInRange
+      elif src.kind in nkFloatLit..nkFloat64Lit and
+          src.floatVal notin
+            firstOrd(targetTyp).toBiggestFloat..lastOrd(targetTyp).toBiggestFloat:
```
I've realized that this isn't entirely correct, due to `high(int64).toBiggestFloat` getting rounded up since it's not representable in `float64`. Is there a way to get the largest representable float that is less-or-equal than some integer?

Araq replied:
Yeah, no idea either, sorry.

GULPF commented:
Rebased and cleaned up the commit mess. To get the correct behavior for conversion of very large floats to int64 I added `nextAfter` and `toBiggestFloatSlice` to the `math` module. If those should not be in the stdlib I can put them somewhere else, but the compiler will need it in multiple places so it needs to reusable.

Araq commented:
@LemonBoy got superseeded by your PRs?

Araq commented:
@narimiran Please check which of these bugs are still open and how to integrate this PR.

narimiran commented:
@GULPF sorry for one unneeded force-push — I was trying to resolve the conflicts locally but something went very wrong with appveyor after a (regular) push containing the commit with the merge, so I took that back and resolved the conflicts here on Github. 

(And now I'm trying to go through the failed testing attempts)

narimiran commented:
Now that devel is successfully merged, there is still one breaking change that should be addressed.

In `tests/statictypes/tstatictypes.nim`, lines 135–139 now fail with `type mismatch: got <N> but expected 'int8'`.

Changing those lines to this:
```nim
block:
  type
    Coord[N: static[int]] = tuple[col, row: range[0'i8 .. (N-1).int8]]
    Point[N: static[int]] = range[0'i16 .. (N * N - 1).int16]
``` 
would make them work, but @Araq confirmed that "`N.int8 - 1` should be of type int8", i.e. everything should work with the existing versions (`N.int8 - 1` and `N.int16 * N.int16 - 1`).

@GULPF, can you make your changes work with the existing test?
